dev/Kozani/KozaniRemoteManager/ConnectionManager.cpp

```diff
@@ -0,0 +1,172 @@
+                    g_module->IncrementObjectCount();
+                    m_isModuleObjectCountAdded = true;
+                }
+            }
+            catch (...)
+            {
+                // Failing to create DVC server is fatal and we should shut down this COM server and let a new instance to be 
+                // started the next COM call comes in. 
+                if (m_isModuleObjectCountAdded)
+                {
+                    // By decementing the object count of this module to remove the count for this object, the object count will
+                    // be reduced to 0 after the current active COM call is finished and the COM server can be shut down. Any new
+                    // COM calls will be routed to a new instance of the COM server.
+                    g_module->DecrementObjectCount();
+                    m_isModuleObjectCountAdded = false;
+                }
+                throw;
+            }
+        }
+        
+        m_dvcServer->SendConnectionAck(connectionId);
+    }
+
+    // Report error from DVC server object. Called from the DVC server object encounting error.
+    void ConnectionManager::ReportDvcServerError(HRESULT hr)
+    {
+        auto lock{ std::unique_lock<std::recursive_mutex>(m_dvcServerLock) };
+
+        LogDebugMessage("DVC server reported error: 0x%x. Will shut down the server and create a new one for the next connection.", hr);
+
+        if (m_dvcServer == nullptr)
+        {
+            // Already deleted previously. Nothing to do.
+            return;
+        }
+
+        WaitForFailedDvcServerDeletionThread();
+
+        m_failedDvcServerToBeDeleted.reset();
+        m_failedDvcServerToBeDeleted.swap(m_dvcServer);
+
+        // Start a separate thread to delete the failed DVC server object as the current thread is from the DVC server object.
+        m_failedDvcServerDeletionThread = std::thread([this] { m_failedDvcServerToBeDeleted.reset(); });
+    }
+
+    void ConnectionManager::WaitForFailedDvcServerDeletionThread()
+    {
+        if (m_failedDvcServerDeletionThread.joinable())
+        {
+            try
+            {
+                m_failedDvcServerDeletionThread.join();
+            }
+            catch (std::invalid_argument)
+            {
+                // std::invalid_argument will be thrown if the thread has exited (no longer joinable). It may happen if the thread is exiting while join() is called.
+                // It is not an issue as we want the thread to exit.
+            }
+        }
+    }
+
+    UINT64 ConnectionManager::GetNewActivityId()
+    {
+        auto lock{ m_newActivityIdLock.lock_exclusive() };
+        return m_newActivityId++;
+    }
+
+    void ConnectionManager::ProcessAppTermination(AppActivationInfo* appActivationInfo)
+    {
+        auto lock{ m_activityMapLock.lock_exclusive() };
+        if (m_closing)
+        {
+            return;
+        }
+
+        UINT64 activityId{ appActivationInfo->activityId };
```

Comment:
I believe using struct for internal code results in more efficient cod in turns runtime performance and dev costs. If a struct need to share across .cpp files then it may make sense to turn it into a class with getters and setters.